### PR TITLE
Fixed #21231 -- Added mechanism to limit request data upload size.

### PR DESCRIPTION
**Pull request by timgraham**
Added settings.DATA_UPLOAD_MAX_MEMORY_SIZE, similar to the existing
FILE_UPLOAD_MAX_MEMORY_SIZE, and raised SuspiciousOperation on
requests that exceed it.

### CODE REVIEW

File: django/http/multipartparser.py [hunk 1]
```diff
@@ -163,15 +168,24 @@ def parse(self):
...
                         try:
                             data = base64.b64decode(raw_data)
                         except _BASE64_DECODE_ERROR:
                             data = raw_data
                     else:
-                        data = field_stream.read()
+                        data = field_stream.read(**read_kwargs)
+
+                    self._data_size += len(data)
+                    if self._max_data_size is not None and field_stream.read(1):
```
Comment (edevil) [inline django/http/multipartparser.py]: When the stream is empty an empty string is returned immediately.

File: django/http/request.py [hunk 2]
```diff
@@ -239,28 +255,23 @@ def _load_post_and_files(self):
...
-                # be explicit about it instead of simply representing an
-                # empty POST
-                self._mark_post_parse_error()
-                raise
-        elif self.META.get('CONTENT_TYPE', '').startswith('application/x-www-form-urlencoded'):
-            self._post, self._files = QueryDict(self.body, encoding=self._encoding), MultiValueDict()
-        else:
-            self._post, self._files = QueryDict('', encoding=self._encoding), MultiValueDict()
+                self._post, self._files = QueryDict(encoding=self._encoding), MultiValueDict()
+        except:
```
Comment (edevil) [inline django/http/request.py]: Corrected in PR #3852.

File: django/http/request.py [hunk 3]
```diff
@@ -204,13 +204,18 @@ def upload_handlers(self, upload_handlers):
...
 
-    def parse_file_upload(self, META, post_data):
+    def parse_file_upload(self):
         """Returns a tuple of (POST QueryDict, FILES MultiValueDict)."""
         self.upload_handlers = ImmutableList(
             self.upload_handlers,
             warning="You cannot alter upload handlers after the upload has been processed."
         )
-        parser = MultiPartParser(META, post_data, self.upload_handlers, self.encoding)
+        # Use already read data, if it exists, otherwise stream the request.
```
Comment (edevil) [inline django/http/request.py]: Removed extra code in PR #3852.

File: django/conf/global_settings.py [hunk 4]
```diff
@@ -305,6 +305,11 @@
 # file system instead of into memory.
 FILE_UPLOAD_MAX_MEMORY_SIZE = 2621440  # i.e. 2.5 MB
 
+# Maximum size in bytes of request data before a SuspiciousOperation will
+# be raised.  This applies to accessing request.body and also to any
+# data in form data requests, excluding file uploads.
+DATA_UPLOAD_MAX_MEMORY_SIZE = 2621440  # i.e. 2.5 MB
```
Comment (edevil) [inline django/conf/global_settings.py]: Servers DO limit the size of the URL including GET parameters.

Maybe this protection is not enough, but the "problems" you mention seem intrinsic to the Python VM and not something we can fix easily here. So, having this protection in place is much better than not having any.

As for making this limit per-view, that seems like a nice-to-have feature. We accept patches. :)

File: docs/ref/settings.txt [hunk 5]
```diff
@@ -854,6 +854,23 @@ TEST_TBLSPACE_TMP
...
+
+.. versionadded:: 1.8
+
+Default: ``2621440`` (i.e. 2.5 MB).
+
+The maximum size (in bytes) that a request body may be before raising a
+:exc:`~django.core.exceptions.SuspiciousOperation` error. This value is applied
+when accessing ``request.body`` or ``request.POST`` and is calculated against
+the total request size excluding any file upload data. You can set it to
+``None`` to disable the check.
```
Comment (tomchristie) [inline docs/ref/settings.txt]: Would it be worth also adding a note recommending that users consider configuring a maximum upload size on their server as well, where possible?  (Not always an option if eg on a managed hosting platform, but more suitable in other cases, with the application level restriction being our general purpose fall-back)

**Pull request closed without merging**